File: ios/NuruNuru/Data/MlsDbKeyStore.swift
import Foundation
import Security

#if NURUNURU_FFI_AVAILABLE
import NuruNuruFFILib
#endif

/// Issue #181: provides the 32-byte SQLCipher key that the Rust core uses to
/// encrypt `<AppSupport>/nurunuru_ndb_mls.sqlite3` on disk.
///
/// Mirror of Android `MlsDbKeyStore.kt`. Two key derivation paths:
///
/// 1. **Internal signer** — caller already holds the nsec hex; we derive
///    deterministically via HKDF-SHA256 (FFI helper). No persistence needed:
///    the key is regenerable from the nsec, so it survives reinstall as long
///    as the user has their nsec backed up.
///
/// 2. **External signer (NIP-46 bunker)** — no nsec is ever exposed to this
///    process. We generate a random 32-byte secret, scope it by pubkey hex,
///    and store it in Keychain with `kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly`.
///    Stronger device-binding, at the cost of being unrecoverable after
///    factory reset (and explicitly NOT iCloud-backed via the
///    `ThisDeviceOnly` accessibility class).
///
/// **Threading**: `getOrCreateExternalKey` is guarded by an in-process
/// `NSLock` so concurrent first-launch threads cannot generate two different
/// random keys and race on the Keychain write (issue #181 B4).
enum MlsDbKeyStore {

    /// App-scope salt for HKDF derivation. Bump suffix if the scheme changes.
    /// Must match Android `MlsDbKeyStore.APP_SALT`.
    static let appSalt = "io.nurunuru.mdk.v1"

    private static let keychainService = "io.nurunuru.app.mls"
    private static let accountPrefixExternalKey = "external_mls_db_key_"

    /// Synchronizes external-key generation across threads (issue #181 B4).
    private static let lock = NSLock()

    // MARK: - Internal signer (HKDF)

    /// Deterministic SQLCipher key for the internal-signer path.
    /// Equivalent to calling `derive_mls_db_key(nsec_bytes, APP_SALT)` in Rust.
    ///
    /// - Parameter secretKeyHex: 64-char nsec hex (no `nsec1` prefix).
    /// - Returns: 32-byte HKDF-SHA256 output as `Data`.
    static func deriveInternalKey(secretKeyHex: String) throws -> Data {
#if NURUNURU_FFI_AVAILABLE
        return try deriveMlsDbKeyFromSecret(secretKeyHex: secretKeyHex, appSalt: appSalt)
#else
        throw MlsDbKeyStoreError.ffiUnavailable
#endif
    }

    // MARK: - External signer (random + Keychain)

    /// Returns the cached 32-byte SQLCipher key for the given pubkey,
    /// generating + persisting a new random one on first call.
    ///
    /// We deliberately use Keychain (NOT UserDefaults) with the most
    /// conservative accessibility class so the key is:
    ///   - never present in iCloud Keychain backups
    ///   - unavailable while the device is locked at first boot
    ///   - wiped on factory reset
    ///
    /// - Parameter pubkeyHex: 64-char lowercase hex pubkey.
    static func getOrCreateExternalKey(pubkeyHex: String) throws -> Data {
        guard pubkeyHex.count == 64 else {
            throw MlsDbKeyStoreError.invalidPubkey
        }
        lock.lock()
        defer { lock.unlock() }

        let account = accountPrefixExternalKey + pubkeyHex.lowercased()

        if let existing = readKeychain(account: account) {
            if existing.count == 32 {
                return existing
            }
            // Wrong-length entry: legacy or corrupt. Delete + regenerate.
            deleteKeychain(account: account)
        }

        var fresh = Data(count: 32)
        let status = fresh.withUnsafeMutableBytes { ptr -> Int32 in
            guard let base = ptr.baseAddress else { return errSecAllocate }
            return Int32(SecRandomCopyBytes(kSecRandomDefault, 32, base))
        }
        guard status == errSecSuccess else {
            throw MlsDbKeyStoreError.randomFailed(status: status)
        }

        // We must NOT return a key that isn't persisted — the encrypted DB
        // we'd create with it would become unrecoverable after process exit.
        try writeKeychain(account: account, data: fresh)
        return fresh
    }

    /// Removes the pubkey-scoped external key. Called from `AuthViewModel.logout()`
    /// *before* `prefs.clear()` so we still have the pubkey to scope by.
    static func clearExternalKey(pubkeyHex: String) {
        lock.lock()
        defer { lock.unlock() }
        let account = accountPrefixExternalKey + pubkeyHex.lowercased()
        deleteKeychain(account: account)
    }

    /// Removes ALL external keys. Used on full wipe.
    static func clearAllExternalKeys() {
        lock.lock()
        defer { lock.unlock() }
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService
        ]
        SecItemDelete(query as CFDictionary)
    }

    // MARK: - Keychain helpers

    private static func readKeychain(account: String) -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        guard status == errSecSuccess, let data = result as? Data else { return nil }
        return data
    }

    private static func writeKeychain(account: String, data: Data) throws {
        // Delete existing entry first to avoid `errSecDuplicateItem`.
        deleteKeychain(account: account)
        let attrs: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: account,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
            kSecValueData as String: data
        ]
        let status = SecItemAdd(attrs as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw MlsDbKeyStoreError.keychainWriteFailed(status: status)
        }
    }

    private static func deleteKeychain(account: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: account
        ]
        SecItemDelete(query as CFDictionary)
    }
}

enum MlsDbKeyStoreError: Error, LocalizedError {
    case ffiUnavailable
    case invalidPubkey
    case randomFailed(status: Int32)
    case keychainWriteFailed(status: Int32)

    var errorDescription: String? {
        switch self {
        case .ffiUnavailable:
            return "MLS FFI is not available (xcframework not linked)"
        case .invalidPubkey:
            return "MLS DB key requires a 64-char hex pubkey"
        case .randomFailed(let status):
            return "SecRandomCopyBytes failed (status=\(status))"
        case .keychainWriteFailed(let status):
            return "Keychain write failed (status=\(status))"
        }
    }
}

File: android/app/src/main/kotlin/io/nurunuru/app/data/MlsDbKeyStore.kt
package io.nurunuru.app.data

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import java.security.SecureRandom
import uniffi.nurunuru.deriveMlsDbKeyFromSecret

/**
 * Issue #181: provides the 32-byte SQLCipher key that the Rust core uses to
 * encrypt `nostrdb_ndb_mls.sqlite3` on disk.
 *
 * Two key derivation paths:
 *
 * 1. **Internal signer** — caller already holds the nsec hex; we derive
 *    deterministically via HKDF-SHA256 (FFI helper). No persistence needed:
 *    the key is regenerable from the nsec, so it survives reinstall as
 *    long as the user has their nsec backed up.
 *
 * 2. **External signer (Amber / NIP-46)** — no nsec is ever exposed to
 *    this process. We generate a random 32-byte secret, scope it by
 *    pubkey hex, and store it in `EncryptedSharedPreferences` backed by
 *    a `MasterKey` rooted in Android Keystore. Stronger device-binding,
 *    at the cost of being unrecoverable after factory reset.
 *
 * **Threading**: `getOrCreateExternalKey` is guarded by an in-process
 * lock (`LOCK`) so concurrent first-launch threads cannot generate two
 * different random keys and race on commit (issue #181 B4).
 */
object MlsDbKeyStore {

    /** App-scope salt for HKDF derivation. Bump suffix if the scheme changes. */
    const val APP_SALT = "io.nurunuru.mdk.v1"

    private const val TAG = "MlsDbKeyStore"
    private const val PREFS_NAME = "nuru_mls_keystore"
    private const val PREF_PREFIX_EXTERNAL_KEY = "external_mls_db_key_"

    /** Synchronizes external-key generation across threads (issue #181 B4). */
    private val LOCK = Any()

    // ─── Internal signer (HKDF) ───────────────────────────────────────

    /**
     * Deterministic SQLCipher key for the internal-signer path.
     * Equivalent to calling `derive_mls_db_key(nsec_bytes, APP_SALT)` in Rust.
     *
     * @param secretKeyHex 64-char nsec hex (no `nsec1` prefix).
     */
    fun deriveInternalKey(secretKeyHex: String): ByteArray {
        return deriveMlsDbKeyFromSecret(secretKeyHex, APP_SALT)
    }

    // ─── External signer (random + keystore) ──────────────────────────

    /**
     * Returns the cached 32-byte SQLCipher key for the given pubkey,
     * generating + persisting a new random one on first call.
     *
     * `commit()` (not `apply()`) is used so the bytes are durable on disk
     * before the caller hands the key to the Rust ctor; an `apply()`-and-
     * crash sequence could leave the prefs without the key while the
     * encrypted DB exists, making the DB unrecoverable.
     *
     * @param pubkeyHex 64-char lowercase hex pubkey.
     */
    fun getOrCreateExternalKey(context: Context, pubkeyHex: String): ByteArray {
        require(pubkeyHex.length == 64) {
            "MlsDbKeyStore.getOrCreateExternalKey: pubkeyHex must be 64-char hex"
        }
        synchronized(LOCK) {
            val prefs = openPrefs(context)
            val prefKey = PREF_PREFIX_EXTERNAL_KEY + pubkeyHex.lowercase()
            val existing = prefs.getString(prefKey, null)
            if (existing != null) {
                val bytes = Base64.decode(existing, Base64.NO_WRAP)
                if (bytes.size == 32) return bytes
                Log.w(TAG, "Stored external MLS key has wrong length ${bytes.size}; regenerating")
            }
            val fresh = ByteArray(32).also { SecureRandom().nextBytes(it) }
            val ok = prefs.edit()
                .putString(prefKey, Base64.encodeToString(fresh, Base64.NO_WRAP))
                .commit()
            if (!ok) {
                // Don't return a key that isn't persisted — the encrypted DB
                // we'd create with it would become unrecoverable.
                throw IllegalStateException(
                    "MlsDbKeyStore: failed to persist external MLS DB key"
                )
            }
            return fresh
        }
    }

    /**
     * Removes the pubkey-scoped external key on logout. Called from
     * AuthViewModel.logout() *before* purging the MLS SQLite file.
     */
    fun clearExternalKey(context: Context, pubkeyHex: String) {
        synchronized(LOCK) {
            val prefs = openPrefs(context)
            val prefKey = PREF_PREFIX_EXTERNAL_KEY + pubkeyHex.lowercase()
            prefs.edit().remove(prefKey).commit()
        }
    }

    /** Removes ALL external keys. Used on full logout / wipe. */
    fun clearAllExternalKeys(context: Context) {
        synchronized(LOCK) {
            openPrefs(context).edit().clear().commit()
        }
    }

    private fun openPrefs(context: Context): SharedPreferences {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        return EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }
}
